### PR TITLE
[build] Put decoders in single decode library

### PR DESCRIPTION
**Pull request by Alexandr-Konovalov**

### CODE REVIEW

**Comment by dmitry-gurulev** [comment]
@Alexandr-Konovalov,

Could u please provide a justification why do we need to proceed this way? For me, another approach -  ([[build] Separate component build to dedicated CMakeLists](https://github.com/Intel-Media-SDK/MediaSDK/pull/1572)) is much better.

**Comment by Alexandr-Konovalov** [comment]
> @Alexandr-Konovalov,
> 
> Could u please provide a justification why do we need to proceed this way? For me, another approach - ([[build] Separate component build to dedicated CMakeLists](https://github.com/Intel-Media-SDK/MediaSDK/pull/1572)) is much better.

@dmitry-gurulev , I think that approach "one component - one CMakeLists.txt" is more "structured", so to say, and dislike multiple unrelated `include_directories`. But @dvrogozh is sure that having single big library is simplifying modifications, if I understand him correctly.

File: _studio/mfx_lib/decode/CMakeLists.txt [hunk 1]
```diff
@@ -20,118 +20,156 @@
...
-foreach( dir ${Decoders} )
-  include_directories( ${MSDK_LIB_ROOT}/decode/${dir}/include )
-  include_directories( ${MSDK_UMC_ROOT}/codec/${dir}_dec/include )
+list( APPEND vdirs
+  h264 h265 mjpeg mpeg2 vc1 vp8 vp9 av1
+  )
+list( APPEND cdirs
+  h265_dec h264_dec mpeg2_dec vp9_dec av1_dec vc1_dec jpeg_dec vc1_common color_space_converter jpeg_common
+  )
+mfx_include_dirs( )
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/CMakeLists.txt]: mfx_include_dirs() is lost here. Put it on the very top as it was before.

File: _studio/mfx_lib/decode/CMakeLists.txt [hunk 2]
```diff
@@ -20,118 +20,156 @@
...
+foreach( dir ${vdirs} )
+  include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/${dir}/include )
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/CMakeLists.txt]: You know, I would prefer to expand foreach loop and put dirs manually. The reason is that potentially mediasdk build is configurable to enable/disable codecs. If we will put dirs manually we will be able to write (in a future):
```
if (MFX_ENABLE_H264_DECODER)
  include(...)
  list(sources APPEND ...)
endif()
```

File: _studio/mfx_lib/decode/CMakeLists.txt [hunk 3]
```diff
@@ -20,118 +20,156 @@
...
-include_directories( ${MSDK_UMC_ROOT}/codec/jpeg_dec/include )
-include_directories( ${MSDK_UMC_ROOT}/codec/jpeg_enc/include )
-include_directories( ${MSDK_UMC_ROOT}/codec/jpeg_common/include )
-include_directories( ${MSDK_UMC_ROOT}/codec/vc1_common/include )
-include_directories( ${MSDK_UMC_ROOT}/codec/color_space_converter/include )
+include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/../vpp/include )
 
-set( sources.plus "" )
+foreach( dir ${cdirs} )
+  include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/../../shared/umc/codec/${dir}/include )
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/CMakeLists.txt]: put includes below definition of UMC_CODECS and reuse it for simplicity

File: _studio/mfx_lib/decode/CMakeLists.txt [hunk 4]
```diff
@@ -20,118 +20,156 @@
...
-make_library( jpeg_common none static )
+set(UMC_CODECS ${CMAKE_CURRENT_SOURCE_DIR}/../../shared/umc/codec/)
 
-### JPEG Encoder
-set( SRC_DIR "${MSDK_UMC_ROOT}/codec/jpeg_enc/src" )
-set( sources "" )
-set( defs "" )
-file( GLOB_RECURSE srcs "${SRC_DIR}/*.cpp" )
-list( APPEND sources ${srcs})
-make_library( jpeg_enc none static )
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/CMakeLists.txt]: jpeg_enc in "decode" cmake:). I feel as if someone tries to hide something:):).

File: _studio/mfx_lib/encode/CMakeLists.txt [hunk 5]
```diff
@@ -24,3 +24,20 @@ foreach( dir ${vdirs} )
...
+    jpeg_common
+    )
+  foreach( dir ${cdirs} )
+    include_directories( ${MSDK_UMC_ROOT}/codec/${dir}/include )
+  endforeach( )
+
+  set( SRC_DIR "${MSDK_UMC_ROOT}/codec/jpeg_enc/src" )
+  set( sources "" )
+  set( defs "" )
+  file( GLOB_RECURSE srcs "${SRC_DIR}/*.cpp" )
```
Comment (dvrogozh) [inline _studio/mfx_lib/encode/CMakeLists.txt]: Advise: don't use glob. Expand.

File: _studio/mfx_lib/encode/CMakeLists.txt [hunk 6]
```diff
@@ -24,3 +24,20 @@ foreach( dir ${vdirs} )
...
+  list( APPEND sources ${srcs})
+  make_library( jpeg_enc none static )
```
Comment (dvrogozh) [inline _studio/mfx_lib/encode/CMakeLists.txt]: I think about encode_sw as a library.

**Comment by dvrogozh** [comment]
@dmitry-gurulev : yeah, that's trade off. Few bigger cmake files are times easier to support as it turns out.

File: _studio/mfx_lib/decode/CMakeLists.txt (hunk 1 above)
Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/decode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/decode/CMakeLists.txt (hunk 2 above)
Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/decode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/decode/CMakeLists.txt (hunk 3 above)
Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/decode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/encode/CMakeLists.txt (hunk 5 above)
Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/encode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/encode/CMakeLists.txt (hunk 6 above)
Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/encode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/decode/CMakeLists.txt [hunk 7]
```diff
@@ -20,118 +20,161 @@
...
+    list(APPEND sources
+        ${UMC_CODECS}/color_space_converter/src/umc_video_processing.cpp
+        ${UMC_CODECS}/color_space_converter/src/umc_color_space_conversion.cpp
+        ${UMC_CODECS}/color_space_converter/src/umc_deinterlacing.cpp
+        )
+endif()
+
+set( sources.plus "" )
+
+set( defs "${API_FLAGS} ${WARNING_FLAGS}" )
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/CMakeLists.txt]: I believe we get rid of API_FLAGS and WARNING_FLAGS here on github. Please, don't introduce again.

Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/decode/CMakeLists.txt]: Done.

File: _studio/mfx_lib/encode/CMakeLists.txt [hunk 8]
```diff
@@ -1,26 +1,33 @@
...
-foreach( dir ${vdirs} )
-  include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/${dir}/include )
-endforeach()
 
-foreach( dir ${cdirs} )
-  include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/../../shared/umc/codec/${dir}/include )
-endforeach( )
+include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/mjpeg/include )
 
 set( defs "${API_FLAGS} ${WARNING_FLAGS}" )
```
Comment (dvrogozh) [inline _studio/mfx_lib/encode/CMakeLists.txt]: @Alexandr-Konovalov : I know that's not your change, but can you, please, remove this line? Really, we don't have these variables on github. Please, just add:
`set( defs "" )`

Comment (Alexandr-Konovalov) [reply _studio/mfx_lib/encode/CMakeLists.txt]: Done.

**Comment by onabiull** [comment]
Could you please rebase to catch up infra changes?

**Pull request merged**